File: Pythoncode/17.电话号码的字母组合.py
class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        d = {}
        d["2"] = ["a", "b", "c"]
        d["3"] = ["d", "e", "f"]
        d["4"] = ["g", "h", "i"]
        d["5"] = ["j", "k", "l"]
        d["6"] = ["m", "n", "o"]
        d["7"] = ["p", "q", "r", "s"]
        d["8"] = ["t", "u", "v"]
        d["9"] = ["w", "x", "y", "z"]
        l = list(digits)
        res = []
        s = ""
        self.helper(l, d, s, res)
        return res

    def helper(self, l, d, s, res):
        if len(l) == 0:
            if len(s) != 0:
                res.append(s)
            return
        n = l[0]
        for c in d[n]:
            news = s[:]
            news += c
            self.helper(l[1:], d, news, res) 
            

File: Cppcode/17.电话号码的字母组合.cpp
#include <string>
#include <vector>

using namespace std;

class Solution {
public:
    vector<vector<char>> n2chars;

    vector<string> letterCombinations(string digits) {
        vector<string> res;
        if (digits == "") return res;
        n2chars = { 
                    {},
                    {},
                    {'a', 'b', 'c'},
                    {'d', 'e', 'f'},
                    {'g', 'h', 'i'},
                    {'j', 'k', 'l'},
                    {'m', 'n', 'o'},
                    {'p', 'q', 'r', 's'},
                    {'t', 'u', 'v'},
                    {'w', 'x', 'y', 'z'}
                  };
        helper(res, digits, "");
        return res;
    }

    void helper(vector<string> &res, string digits, string cur_string)
    {
        if (digits == "")
        {
            res.push_back(cur_string);
        }
        else
        {
            for (char c : n2chars[stoi(string(1, digits.front()))])
            {
                helper(res, digits.substr(1), cur_string + c);
            }
        }
    }
};